### DISCUSSION TITLE
Nginx and PHP performance optimisation.

### DISCUSSION THREAD

**Discussion by cihantas** [dokku/dokku#4408] [Q&A] [answered]
Hi,

We're running Dokku on a DigitalOcean droplet. We recently had a lot of unexpected growth in traffic, which sadly ended in a bunch of 502 and 499s and `upstream timed out (110: Connection timed out)` because all Nginx connections appeared to be in use.

The system limits for file descriptors seem to above the defaults. `ulimit -Hn` returns `1048576` both on host and within container.

Entering the web container for a PHP app and checking `/app/.heroku/php/etc/nginx/nginx.conf` showed `worker_process 1` and `worker_connections 1024`. Both aren't high enough. This is not meant to criticise Dokku. We're well aware that these are just sane defaults and that these values are depending on the use-case/resources.

**What's the best way to customise the configuration that lie outside of `http {}` block? Do we have to fork the buildpack or is there a simple alternative to change these values?**

Are there any other Nginx / PHP-specific performance optimisation you can recommend? PHP-FPM has toggles like `process.max 0` `rlimit_files` which are already set to respect the system value or just have no limit.  

Would appreciate to learn from you. Thanks in advance.

**Comment by josegonzalez** [comment, marked as answer]
You can use a custom nginx.conf for the php buildpack, as shown here: https://devcenter.heroku.com/articles/custom-php-settings#nginx

For anything outside that server block, I think you'll need to fork and customize the buildpack.

**Comment by rezart** [comment]
I am running into a similar issue. The defaults might be a bit low for nginx, especially considering today's modern hardware. Anyways, the issue for me was with worker_connections inside the events directive. I edited the main nginx.conf file inside of `/etc/nginx/nginx.conf` setting the worker_connections to the same number as my `ulimit -n` number and reloaded the nginx service with `sudo service nginx reload`. 

The traffic has died down for the time being so hard for me to test if this worked. @josegonzalez any thoughts if this would do the trick?

**Reply by cihantas** [reply]
@rezart I made some progress by forking [Heroku's Buildpack](https://github.com/heroku/heroku-buildpack-php) and changing the `worker_connections` value in [main.conf](https://github.com/heroku/heroku-buildpack-php/blob/main/conf/nginx/main.conf).

**Reply by josegonzalez** [reply]
Optimizing nginx is a bit of magic, so there won't be a silver bullet for every config. I definitely suggest trying to load test your setup to ensure optimal performance for the traffic you are expecting.
